### ISSUE TITLE
G1 feet sink into the ground during motion lib playback

### ISSUE THREAD

**Issue by jaraujo98** [NVlabs/ProtoMotions#167]
My reference motion data has some floating in it (I use my own retargeting logic, which fixes floating by using the original Unitree collision geometry, which are four small spheres at the bottom of each ankle link)

https://github.com/user-attachments/assets/33aae89d-fe4a-46c2-90a5-4aef2e6c123a

When training the Mimic controller I found that the agent learned to jump all the time. I changed to the BeyondMimic reward function since it does not have this problem, and managed to train a Mimic policy that looks good

https://github.com/user-attachments/assets/61dc08b5-2900-49f3-bb07-b449e06a2b95

You can see the robot falling down at the start, but after that it works well. I decided to rollout my policy on my motion library to get physically accurate reference motion. However, when I replay it using the `examples/motion_libs_visualizer.py` script, the feet sink into the ground.

https://github.com/user-attachments/assets/9652f246-1856-486f-bc3b-7c28a76dfc94

I am inclined to think that my motion data is good and there is a bug with the `motion_libs_visualizer.py` script, but I would like to check. My attempts at changing the height of the floor on that script by editing this line

https://github.com/NVlabs/ProtoMotions/blob/56e49e7f4443fb06085fae9a2eed825f0d3e85b7/examples/motion_libs_visualizer.py#L289

have failed, so I really don't know what is going on here.

**Comment by tesslerc** [comment]
For training the G1 did you use maximal coordinates or reduced (like in `mimic/domain_rand.py`)?
We typically don't see much jumping with the G1, but we also provide reduced coordinate obs. Maybe there's a difference there?

[edit] Maybe better if you could share the config you used that caused issues and the new config you train with that solved this? [/edit]

Regarding recording -- how are you recording the motion? I'd like to try the same thing on our end with G1 and SMPL humanoids and see if this also happens or if there's some other underlying issue.

**Comment by jaraujo98** [comment]
For recording, I'm using the `L` key shortcut. It stores what I see in the viewer (namely the feet sinking under the ground plane).

For my experiments, I used the default mimic tracking configuration:

```bash
python protomotions/train_agent.py \
    --robot-name g1 \
    --simulator isaaclab \
    --experiment-path examples/experiments/mimic/mlp.py \
    --experiment-name ... \
    --motion-file ... \
    --num-envs 8192 \
    --batch-size 8192 \
    --use-wandb
```

which uses maximal coordinates (`mimic/mlp_domain_rand.py` seems to also use maximal? At least the released version.). Here's an example of the hopping when rolling out the policy (a fairly exaggerated one)

https://github.com/user-attachments/assets/97d268a0-3ff5-474a-a97d-5be75bbc2d2a

The way I fixed this was by modifying the reward configuration (all the reward functions come from the BeyondMimic public implementation)

<details>

<summary>BeyondMimic reward configuration</summary>

```python
        reward_config = {
        # BeyondMimic rewards
        "motion_global_anchor_pos": RewardComponentConfig(
            function=motion_global_anchor_position_error_exp,
            variables={
                "x": "current_state.rigid_body_pos[:, 0, :]",
                "ref_x": "ref_state.rigid_body_pos[:, 0, :]",
                "std": "0.3",
            },
            weight=0.5,
        ),
        "motion_global_anchor_ori": RewardComponentConfig(
            function=motion_global_anchor_orientation_error_exp,
            variables={
                "x": "current_state.rigid_body_rot[:, 0, :]",
                "ref_x": "ref_state.rigid_body_rot[:, 0, :]",
                "std": "0.4",
            },
            weight=0.5,
        ),
        "motion_body_pos": RewardComponentConfig(
            function=motion_relative_body_position_error_exp,
            variables={
                "current_state": "current_state",
                "ref_state": "ref_state",
                "std": "0.3",
            },
            indices_subset=KEY_LINKS,
            weight=1.0,
        ),
        "motion_body_ori": RewardComponentConfig(
            function=motion_relative_body_orientation_error_exp,
            variables={
                "current_state": "current_state",
                "ref_state": "ref_state",
                "std": "0.4",
            },
            indices_subset=KEY_LINKS,
            weight=1.0,
        ),
        "motion_body_lin_vel": RewardComponentConfig(
            function=motion_global_body_linear_velocity_error_exp,
            variables={
                "x": "current_state.rigid_body_vel",
                "ref_x": "ref_state.rigid_body_vel",
                "std": "1.0",
            },
            weight=1.0,
        ),
        "motion_body_ang_vel": RewardComponentConfig(
            function=motion_global_body_angular_velocity_error_exp,
            variables={
                "x": "current_state.rigid_body_ang_vel",
                "ref_x": "ref_state.rigid_body_ang_vel",
                "std": "3.14",
            },
            weight=1.0,
        ),
        # Regularization rewards
        "action_smoothness": RewardComponentConfig(
            function=action_rate_l2,
            variables={
                "x": "current_actions - previous_actions",
            },
            weight=-0.01,
        ),
        "joint_limit": RewardComponentConfig(
            function=joint_pos_limits,
            variables={
                "joint_pos": "current_state.dof_pos",
                "soft_dof_limits_lower": "soft_dof_limits_lower",
                "soft_dof_limits_upper": "soft_dof_limits_upper",
            },
            weight=-10.0,
        ),
    }
```
</details>

Everything else remains the same. I was not so familiar with the codebase when I did this change, but, looking at it now, my guess is that the hopping comes from the terms `gt_rew` (absolute difference in link position) and `rh_rew` (root height difference). Are you suggesting I try changing these terms to be functions of differences in reduced coordinates rather than maximal ones?

**Comment by tesslerc** [comment]
You'll see that the `domain_rand` experiment passes reduced coordinates to the actor and maximal coordinates to the critic.
https://github.com/NVlabs/ProtoMotions/blob/56e49e7f4443fb06085fae9a2eed825f0d3e85b7/examples/experiments/mimic/mlp_domain_rand.py#L244

We'll release an update this week to make this workflow simpler (full transition to manager-based).

Regardless, if your motion has floating issues then indeed it may lead to artifacts as our reward requires stricter tracking.
Especially since our reward looks at all joints and not just end-effectors.

----

With respect to the motion visualizer.

I recorded the following motion in Newton on our trained policy.

https://github.com/user-attachments/assets/22c1d318-c00d-42eb-8492-acd99b579edd

I then replayed it with motion visualizer with IsaacLab, without changing any parameters.

https://github.com/user-attachments/assets/30e1f748-76eb-4cf1-8fcc-9090754376c0

It seems to be correct?

I've updated the recording as follows (this will be updated in the code in the upcoming release):
<details>

```
robot_state = self.get_robot_state()
self._recorded_motion["gts"].append(robot_state.rigid_body_pos)
self._recorded_motion["grs"].append(robot_state.rigid_body_rot)
if robot_state.rigid_body_vel is not None:
    self._recorded_motion["gvs"].append(robot_state.rigid_body_vel)
if robot_state.rigid_body_ang_vel is not None:
    self._recorded_motion["gavs"].append(robot_state.rigid_body_ang_vel)
if robot_state.dof_pos is not None:
    self._recorded_motion["dps"].append(robot_state.dof_pos)
if robot_state.dof_vel is not None:
    self._recorded_motion["dvs"].append(robot_state.dof_vel)
if robot_state.rigid_body_contacts is not None:
    self._recorded_motion["contacts"].append(robot_state.rigid_body_contacts)
```

then to save the recording
```
# Save the recorded motion as a .motion file
motion_data = build_motion_data(
    self._recorded_motion,
    fps=30,  # Video recording FPS
    num_dof=self._num_dof,
)
motion_file_path = f"{self._curr_user_recording_name}.motion"
torch.save(motion_data, motion_file_path)
print(f"Motion saved to {motion_file_path}")
self._recorded_motion = None
```

and the helper function is
```
def build_motion_data(
    recorded_motion: Dict[str, List[torch.Tensor]],
    fps: int,
    num_dof: int = 0,
) -> Dict[str, Any]:
    """
    Build a .motion file compatible data dictionary from recorded motion.

    The .motion format uses RobotState field names directly (rigid_body_pos,
    rigid_body_rot, etc.) and is loaded by MotionLib via RobotState.from_dict().

    Args:
        recorded_motion: Dictionary containing lists of tensors for each field.
            Expected keys: gts, grs, gvs, gavs, dps, dvs, contacts
            (using MotionLib naming convention)
        fps: Frames per second of the recorded motion.
        num_dof: Number of DOFs for the robot (used to create placeholder tensors
            if dof data wasn't recorded).

    Returns:
        Dictionary compatible with .motion format (RobotState field names + fps).
    """
    from protomotions.simulator.base_simulator.simulator_state import StateConversion

    # Mapping from recorded field names to RobotState field names
    field_mapping = {
        "gts": "rigid_body_pos",
        "grs": "rigid_body_rot",
        "gvs": "rigid_body_vel",
        "gavs": "rigid_body_ang_vel",
        "dps": "dof_pos",
        "dvs": "dof_vel",
        "contacts": "rigid_body_contacts",
    }

    motion_data: Dict[str, Any] = {
        "fps": fps,
        "state_conversion": StateConversion.COMMON,
    }

    # Concatenate recorded frames for each field
    for field_name, frame_list in recorded_motion.items():
        if len(frame_list) > 0 and field_name in field_mapping:
            # Concatenate along dim 0 (time dimension)
            # Each frame has shape [num_envs, ...], we take only env 0
            stacked = torch.stack(
                [frame[0:1] for frame in frame_list], dim=0
            ).squeeze(1)
            motion_data[field_mapping[field_name]] = stacked

    num_frames = motion_data["rigid_body_pos"].shape[0]
    num_bodies = motion_data["rigid_body_pos"].shape[1]
    device = motion_data["rigid_body_pos"].device
    dtype = motion_data["rigid_body_pos"].dtype

    # Ensure all required fields exist (create zero tensors for missing fields)
    if "rigid_body_vel" not in motion_data:
        motion_data["rigid_body_vel"] = torch.zeros(
            (num_frames, num_bodies, 3), device=device, dtype=dtype
        )
    if "rigid_body_ang_vel" not in motion_data:
        motion_data["rigid_body_ang_vel"] = torch.zeros(
            (num_frames, num_bodies, 3), device=device, dtype=dtype
        )
    if "dof_pos" not in motion_data:
        motion_data["dof_pos"] = torch.zeros(
            (num_frames, num_dof), device=device, dtype=dtype
        )
    if "dof_vel" not in motion_data:
        motion_data["dof_vel"] = torch.zeros(
            (num_frames, num_dof), device=device, dtype=dtype
        )
    if "rigid_body_contacts" not in motion_data:
        motion_data["rigid_body_contacts"] = torch.zeros(
            (num_frames, num_bodies), device=device, dtype=torch.bool
        )
    else:
        # Ensure contacts are boolean type
        motion_data["rigid_body_contacts"] = motion_data["rigid_body_contacts"].bool()

    return motion_data
```
</details>

**Comment by jaraujo98** [comment]
It does look correct. Also, I just realized I misunderstood your question. The way I am recording the rollouts is using the `inference_agent.py` script with `--full-eval`, which calls [this function](https://github.com/NVlabs/ProtoMotions/blob/56e49e7f4443fb06085fae9a2eed825f0d3e85b7/protomotions/agents/evaluators/mimic_evaluator.py#L534) at the end.


I'll try the new release once it comes out, and see if the problem is solved. 😄

**Comment by jyf588** [comment]
To see whether it's a bug on the visualizer side or the data side, maybe we can print out the feet locations from the recorded motionLib, and during inference_agent.py simulation, and see if it is actually sinking?

**Comment by jaraujo98** [comment]
@jyf588 thank you for the suggestion! I have been busy with setting up experiments and haven't had time to look into this issue, but I will check again once my experiments are running (hopefully soon)!

In the meantime, I may have found something. On respawn there is a height offset controlled by `BaseEnv.config.ref_respawn_offset` (`5cm` by default). I can imagine this being useful to avoid ground penetrations on reset. However, when the environment computes the `extras` dict, it subtracts that offset from the motion.

https://github.com/NVlabs/ProtoMotions/blob/56e49e7f4443fb06085fae9a2eed825f0d3e85b7/protomotions/envs/base_env/env.py#L507-L510

The evaluator then copies these `raw` values and stores them in the motion lib. I'm guessing this accounts for both my robot falling down at the start of every rollout (my retargeter fixes my motion to make sure the height of the first frame is correct), and then the offset seen in the replay. I will look into the feet locations on the recorded MotionLib to see if they are all going `5cm` under where they should be.

**Comment by jaraujo98** [comment]
I set `env.ref_respawn_offset=0.0` in the overrides of the inference script and generated the motion file again. The feet are no longer sinking under the ground. I also plotted a histogram of the feet height for both rolled out motion libraries (dark blue is with `env.ref_respawn_offset=0.05`, and light blue is `env.ref_respawn_offset=0.0`).

<img width="365" height="326" alt="Image" src="https://github.com/user-attachments/assets/e76edd85-a337-456d-bd04-b93eb73f1623" />

So this offset seems to be the root cause of the problem.

**Comment by jyf588** [comment]
Thanks for taking a look into this! Yes I see what you mean now - removing L508 is a better design when logging the metrics, and should fix the viz issue when saving replay. We will remove this line in the upcoming release.

btw why would your robot "fall down at the start of every rollout"? Policy should be able to handling spawning it 5cm higher? Did you mean you train it without the offset but test it with the 5cm?

**Comment by jaraujo98** [comment]
Maybe `ref_respawn_offset` could be removed entirely? Penetrations with the ground in the reference motion are a big source of problems (when you reset your simulation to a state that has a lot of it), but they can be solved at the retargeting stage (which is what I do) or even when building the motion library (I've thought about opening a PR for it 😄).

"fall down at the start of every rollout" was a poor choice of words 😅 What I meant was that the robot starts floating in the air (you can see it in the videos from the original issue), and then "falls" until the feet hit the ground. The policy handles everything correctly all the time. The policies I trained up until now use the default 5cm offset, which is also used at inference time. I have yet to train policies with that offset disabled.

Looking forward to the upcoming release!

**Comment by tesslerc** [comment]
@jaraujo98 that's true. Our retargeting is currently assuming the joint positions which isn't as precise. What you're suggesting will require computing the per-frame minimal height of the actual character mesh.
We're very interested in adding this! but haven't done so yet. If you have this, very happy to accept a PR and integrate.

The second thing to keep in mind is terrains. With terrains, a slope might lead to penetrations since computing the terrain height for each mesh vertex is computationally heavy.
Spawning with a small offset shouldn't cause much harm, as the character lands within 2-3 frames and it helps overcome small imperfections.

TL;DR: I do think spawn height offset is useful for avoiding unwanted errors and we should leave a small default value to ensure experiments don't start failing and users can't understand why.
But at the same time I agree that improving the retargeting process to be more precise + showing power users how to disable this is a net positive addition!

**Comment by jaraujo98** [comment]
@tesslerc and @jyf588 thank you for the support and the discussion! I think we have narrowed down the cause to the height offset being subtracted in the wrong place, which seems to be gone in the new release. I had not thought about the scenario with terrains, so I stand corrected about removing the height offset :-) I'm currently sprinting to finish a project in the coming weeks, but once I'm done with that I'll revisit some possible contributions I think I can make to this project :-) In the meantime, I'll close this issue.
